File: packages/urdu-js/src/transliteration/to-roman.ts
import { URDU_TO_ROMAN } from './roman-map.js'

export function toRoman(text: string): string {
  if (!text) return text
  let result = '', i = 0
  while (i < text.length) {
    let matched = false
    for (const [urdu, roman] of URDU_TO_ROMAN) {
      if (text.startsWith(urdu, i)) {
        result += roman; i += urdu.length; matched = true; break
      }
    }
    if (!matched) { result += text[i]!; i++ }
  }
  return result
}

File: packages/urdu-dart/lib/src/transliteration/to_roman.dart
import 'roman_map.dart';

/// Transliterates Urdu text into Roman Urdu based on a mapping table.
String toRoman(String text) {
  if (text.isEmpty) return text;
  final buf = StringBuffer();
  var i = 0;
  while (i < text.length) {
    var matched = false;
    for (final entry in urduToRoman) {
      final urdu = entry.$1;
      final roman = entry.$2;
      if (text.startsWith(urdu, i)) {
        buf.write(roman);
        i += urdu.length;
        matched = true;
        break;
      }
    }
    if (!matched) {
      buf.write(text[i]);
      i++;
    }
  }
  return buf.toString();
}
